pkg/s3/s3.go

```diff
@@ -0,0 +1,94 @@
+package s3
+
+import (
+	"fmt"
+	"math/rand"
+	"reflect"
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+type User struct {
+	Name            string
+	AccessKeyID     string
+	AccessSecretKey string
+}
+
+type Bucket struct {
+	Parameters map[string]string
+	Accesses   map[string]*User
+}
+
+// FakeS3Client is a reference implementation S3 client
+// that use k-v store as a bucket.
+type FakeS3Client struct {
+	Buckets map[string]*Bucket
+}
+
+// NewClient creates new S3 client.
+func NewClient() *FakeS3Client {
+	return &FakeS3Client{
+		Buckets: map[string]*Bucket{},
+	}
+}
+
+// genKey generates random string of lenght n.
+func genKey(n int) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = charset[rand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+// CreateBucket creates a bucket.
+func (s *FakeS3Client) CreateBucket(name string, parameters map[string]string) error {
+	s.Buckets[name] = &Bucket{
+		Parameters: parameters,
+		Accesses:   map[string]*User{},
+	}
+
+	return nil
+}
+
+// BucketExists checks if bucket already exists.
+func (s *FakeS3Client) BucketExists(name string) bool {
+	_, ok := s.Buckets[name]
+	return ok
+}
+
+// IsBucketEqual check equality with new bucket.
+func (s *FakeS3Client) IsBucketEqual(name string, parameters map[string]string) bool {
+	return reflect.DeepEqual(s.Buckets[name].Parameters, parameters)
+}
+
+// DeleteBucket deletes a bucket.
+func (s *FakeS3Client) DeleteBucket(name string) {
+	delete(s.Buckets, name)
+}
```

Comment:
Since this is a sample driver, I think that this can be as-is if we want. 

But I am curious whether we should consider the accesses here as well. 

IIRC, in S3 users are tracked separately from buckets. In this scheme, accesses are a property of buckets, and deleting a bucket will also delete accesses. Should we have this return an error if accesses exist, presumably keeping the bucket in use? I'm curious about others' thoughts. @shanduur @presztak ?

Reply:
I would like to see the relationship inverted, so Access having a reference by ID to the bucket. This will be more in line with how S3 OSP are handling Accesses and Buckets.

Reply:
@shanduur so then we need to keep two maps? One to track buckets and the second one to track accesses with relation to buckets? Am I right?

Reply:
@presztak exactly!

Reply:
Fixed.